### PR TITLE
BigQueryIO: enable users to control the SQL dialect

### PR DESCRIPTION
**Pull request by dhalperi**
BigQuery 2.0 adds a new dialect of SQL, Standard SQL, and we need to
enable passing a specific option to BigQuery's APIs in order to support
it.

### CODE REVIEW

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java [hunk 1]
```diff
@@ -523,6 +549,12 @@ public void validate(PInput input) {
           } else if (query != null && flattenResults == null) {
             throw new IllegalStateException("Invalid BigQuery read operation. Specifies a"
                 + " query without a result flattening preference");
+          } else if (table != null && useLegacySql != null) {
```
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: This follows the pattern of flattenResults.

However, I think we can move the validation of useLegacySql to usingStandardSql(). For example, throws if users do BigQuery.Read.from(table).usingStandardSql().

Also, looks to me flattenResults is not configurable, and can be removed in a separate PR.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java [hunk 2]
```diff
@@ -460,17 +467,23 @@ public Bound from(TableReference table) {
...
+            testBigQueryServices);
       }
 
       /**
        * Disable table validation.
        */
       public Bound withoutValidation() {
-        return new Bound(name, query, jsonTableRef, false, flattenResults, testBigQueryServices);
+        return new Bound(
+            name, query, jsonTableRef, false, flattenResults, useLegacySql, testBigQueryServices);
```
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: false /* validate */

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java [hunk 3]
```diff
@@ -481,12 +494,25 @@ public Bound withoutValidation() {
        * from a table will cause an error during validation.
        */
       public Bound withoutResultFlattening() {
-        return new Bound(name, query, jsonTableRef, validate, false, testBigQueryServices);
+        return new Bound(
+            name, query, jsonTableRef, validate, false, useLegacySql, testBigQueryServices);
```
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: false /* flattenResults*/

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java [hunk 4]
```diff
@@ -481,12 +494,25 @@ public Bound withoutValidation() {
...
+
+      /**
+       * Enables BigQuery's Standard SQL dialect when reading from a query.
+       *
+       * <p>Only valid when a query is used ({@link #fromQuery}). Setting this option when reading
+       * from a table will cause an error during validation.
+       */
+      public Bound usingStandardSql() {
+        return new Bound(
+            name, query, jsonTableRef, validate, flattenResults, false, testBigQueryServices);
```
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: false /* useLegacySql */

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java (hunk 1 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: I took my comment back, I see users could do BigQuery.Read.named().usingStandardSql().from(table) or even BigQuery.Read.fromQuery(query).usingStandardSql().from(table).

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java [hunk 5]
```diff
@@ -534,16 +566,20 @@ public void validate(PInput input) {
...
           if (query != null) {
-            dryRunQuery(bqOptions, query);
+            dryRunQuery(bqOptions, query, useLegacySql);
           }
         }
       }
 
-      private static void dryRunQuery(BigQueryOptions options, String query) {
+      private static void dryRunQuery(
+          BigQueryOptions options, String query, @Nullable Boolean useLegacySql) {
```
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: useLegacySql should not be null for query

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/io/BigQueryIOTest.java [hunk 6]
```diff
@@ -575,6 +576,25 @@ public void testBuildSourceWithTableAndFlatten() {
...
+    thrown.expectMessage(
+        "Invalid BigQuery read operation. Specifies a"
+              + " table with a SQL dialect preference, which is not configurable");
+    p.apply(
+        BigQueryIO.Read.named("ReadMyTable")
+            .from("foo.com:project:somedataset.sometable")
+            .usingStandardSql());
+    p.run();
+  }
+
```
Comment (peihe) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/io/BigQueryIOTest.java]: What is the plan to test the success case?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java (hunk 5 above)
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: No, it definitely can -- it's null if the user has not set it. Look ~10 lines up

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/io/BigQueryIOTest.java (hunk 6 above)
Comment (dhalperi) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/io/BigQueryIOTest.java]: Not sure what you're asking.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java (hunk 5 above)
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: You're right, we enforce this property in the constructor. PTAL

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java [hunk 7]
```diff
@@ -481,69 +495,93 @@ public Bound withoutValidation() {
...
+              "Invalid BigQueryIO.Read: Specifies a table with a SQL dialect"
+                  + " preference, which only applies to queries");
+
+          // Check for source table presence for early failure notification.
+
+          verifyDatasetPresence(bqOptions, table);
+          verifyTablePresence(bqOptions, table);
+        } else /* query != null */ {
+          checkState(flattenResults != null, "flattenResults should not be null if query is set");
+          checkState(useLegacySql != null, "useLegacySql should not be null if query is set");
```
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: checkNotNull for both?

Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: I want to throw `IllegalStateException` not `NullPointerException`
